README.md
# TYPO3 Extension hide_used_content

[![Latest Stable Version](https://img.shields.io/packagist/v/ichhabrecht/hide-used-content.svg)](https://packagist.org/packages/ichhabrecht/hide-used-content)
[![Build Status](https://img.shields.io/travis/IchHabRecht/hide_used_content/master.svg)](https://travis-ci.org/IchHabRecht/hide_used_content)
[![StyleCI](https://styleci.io/repos/155286928/shield?branch=master)](https://styleci.io/repos/155286928)

The extension removes content elements that have an own colPos configuration (e.g. nested content elements) from the
"Unused" column and hides wrong "Unused elements detected on this page" notification. 


ChangeLog
2023-08-31         [RELEASE] Release of hide_used_content 1.1.0 (Nicole Cordes)
2021-09-27 1c4255b Add compatibility to TYPO3 11.5 (Marco Kuprat)
2023-08-31 7f44477 [TASK] Update supporting files to EXT:my_skeleton 0.7.2 (Nicole Cordes)
2020-10-25 7bf8e07 [TASK] Update supporting files to EXT:my_skeleton 0.5.0 (Nicole Cordes)

2020-06-25         [RELEASE] Release of hide_used_content 1.0.0 (Nicole Cordes)
2020-06-25 1d8bdfc [TASK] Update supporting files to EXT:my_skeleton 0.3.2 (Nicole Cordes)
2020-06-25 0210a68 [TASK] Declare TYPO3 10 compatibility (Ludwig Rafelsberger)

2018-10-30         [RELEASE] Release of hide_used_content 0.2.0 (Nicole Cordes)
2018-10-30 d45818c [FEATURE] Use RelationHandler to evaluate relations more accurate (Nicole Cordes)

2018-10-30         [RELEASE] Release of hide_used_content 0.1.0 (Nicole Cordes)
2018-10-30 0f6741c [FEATURE] Add basic PageLayoutViewHook to remove used elements (Nicole Cordes)
2018-10-30 47cf9c9 [FEATURE] Add slot for TCA processing (Nicole Cordes)
2018-10-30 c826f49 [TASK] Add basic extension files (Nicole Cordes)


Classes/Cache/CacheManager.php
<?php

declare(strict_types=1);

namespace IchHabRecht\HideUsedContent\Cache;

use TYPO3\CMS\Core\Cache\Frontend\PhpFrontend;
use TYPO3\CMS\Core\Utility\GeneralUtility;

class CacheManager
{
    /**
     * @var PhpFrontend
     */
    protected $cache;

    protected $cacheIdentifier = 'tx_hideusedcontent';

    public function __construct(PhpFrontend $cache = null)
    {
        $this->cache = $cache ?: GeneralUtility::makeInstance(\TYPO3\CMS\Core\Cache\CacheManager::class)->getCache('cache_core');
    }

    public function get()
    {
        $cacheData = $this->cache->require($this->cacheIdentifier);

        return $cacheData['configuration'] ?? [];
    }

    public function set(array $configuration)
    {
        $this->cache->set(
            $this->cacheIdentifier,
            'return '
            . var_export(
                [
                    'configuration' => $configuration,
                ],
                true
            )
            . ';'
        );
    }
}


Classes/Slot/TcaColPosSlot.php
<?php

declare(strict_types=1);

namespace IchHabRecht\HideUsedContent\Slot;

use IchHabRecht\HideUsedContent\Cache\CacheManager;
use TYPO3\CMS\Core\Utility\GeneralUtility;

class TcaColPosSlot
{
    /**
     * @var CacheManager
     */
    protected $cacheManager;

    public function __construct(CacheManager $cache = null)
    {
        $this->cacheManager = $cache ?: GeneralUtility::makeInstance(CacheManager::class);
    }

    public function initializeColPosCache($tca): array
    {
        $configuration = [];

        foreach ($tca as $table => $tableConfiguration) {
            if (empty($tableConfiguration['columns'])) {
                continue;
            }

            foreach ($tableConfiguration['columns'] as $field => $fieldConfiguration) {
                if (!$this->isInlineField($fieldConfiguration)
                    || !$this->isValidForeignTable($fieldConfiguration)
                    || !$this->hasColPosConfiguration($fieldConfiguration)
                ) {
                    continue;
                }

                $colPos = (int)$fieldConfiguration['config']['overrideChildTca']['columns']['colPos']['config']['default'];
                if (!isset($configuration[$colPos][$table])) {
                    $configuration[$colPos][$table] = [];
                }
                $configuration[$colPos][$table][] = $field;
            }
        }

        $this->cacheManager->set($configuration);

        return [$tca];
    }

    protected function isInlineField(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['type'])
            && $fieldConfiguration['config']['type'] === 'inline';
    }

    protected function isValidForeignTable(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['foreign_table'])
            && !empty($fieldConfiguration['config']['foreign_field'])
            && $fieldConfiguration['config']['foreign_table'] === 'tt_content';
    }

    protected function hasColPosConfiguration(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['overrideChildTca']['columns']['colPos']['config']['default']);
    }
}


Classes/Hooks/PageLayoutViewHook.php
<?php

declare(strict_types=1);

namespace IchHabRecht\HideUsedContent\Hooks;

use IchHabRecht\HideUsedContent\Cache\CacheManager;
use TYPO3\CMS\Core\Database\RelationHandler;
use TYPO3\CMS\Core\Utility\GeneralUtility;

class PageLayoutViewHook
{
    /**
     * @var array
     */
    protected $columnConfiguration;

    public function __construct(CacheManager $cacheManager = null)
    {
        if ($cacheManager === null) {
            $cacheManager = GeneralUtility::makeInstance(CacheManager::class);
        }

        $this->columnConfiguration = $cacheManager->get();
    }

    public function hideUsedContent(array $parameter): bool
    {
        if ($parameter['used']) {
            return true;
        }

        $record = $parameter['record'];
        $colPos = (int)$record['colPos'];

        foreach ($this->columnConfiguration[$colPos] ?? [] as $table => $fieldArray) {
            foreach ($fieldArray as $field) {
                $fieldConfiguration = $GLOBALS['TCA'][$table]['columns'][$field]['config'];
                if (empty($record[$fieldConfiguration['foreign_field']])) {
                    continue;
                }

                $relationHandler = GeneralUtility::makeInstance(RelationHandler::class);
                $relationHandler->start(
                    '',
                    $fieldConfiguration['foreign_table'],
                    $fieldConfiguration['MM'] ?? '',
                    $record[$fieldConfiguration['foreign_field']],
                    $table,
                    $fieldConfiguration
                );
                $valueArray = $relationHandler->getValueArray();
                if (in_array($record['uid'], $valueArray)) {
                    return true;
                }
            }
        }

        return false;
    }
}


Classes/Listener/TcaColPosListener.php
<?php

declare(strict_types=1);

namespace IchHabRecht\HideUsedContent\Listener;

use IchHabRecht\HideUsedContent\Cache\CacheManager;
use TYPO3\CMS\Core\Configuration\Event\AfterTcaCompilationEvent;
use TYPO3\CMS\Core\Utility\GeneralUtility;

class TcaColPosListener
{
    /**
     * @var CacheManager
     */
    protected $cacheManager;

    public function __construct(CacheManager $cache = null)
    {
        $this->cacheManager = $cache ?: GeneralUtility::makeInstance(CacheManager::class);
    }

    public function __invoke(AfterTcaCompilationEvent $event)
    {
        $this->initializeColPosCache($event->getTca());
    }

    public function initializeColPosCache($tca): array
    {
        $configuration = [];

        foreach ($tca as $table => $tableConfiguration) {
            if (empty($tableConfiguration['columns'])) {
                continue;
            }

            foreach ($tableConfiguration['columns'] as $field => $fieldConfiguration) {
                if (!$this->isInlineField($fieldConfiguration)
                    || !$this->isValidForeignTable($fieldConfiguration)
                    || !$this->hasColPosConfiguration($fieldConfiguration)
                ) {
                    continue;
                }

                $colPos = (int)$fieldConfiguration['config']['overrideChildTca']['columns']['colPos']['config']['default'];
                if (!isset($configuration[$colPos][$table])) {
                    $configuration[$colPos][$table] = [];
                }
                $configuration[$colPos][$table][] = $field;
            }
        }

        $this->cacheManager->set($configuration);

        return [$tca];
    }

    protected function isInlineField(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['type'])
            && $fieldConfiguration['config']['type'] === 'inline';
    }

    protected function isValidForeignTable(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['foreign_table'])
            && !empty($fieldConfiguration['config']['foreign_field'])
            && $fieldConfiguration['config']['foreign_table'] === 'tt_content';
    }

    protected function hasColPosConfiguration(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['overrideChildTca']['columns']['colPos']['config']['default']);
    }
}


Classes/EventListener/TcaColPosEventListener.php
<?php

declare(strict_types=1);

namespace IchHabRecht\HideUsedContent\EventListener;

use IchHabRecht\HideUsedContent\Cache\CacheManager;
use TYPO3\CMS\Core\Configuration\Event\AfterTcaCompilationEvent;
use TYPO3\CMS\Core\Utility\GeneralUtility;

class TcaColPosEventListener
{
    /**
     * @var CacheManager
     */
    protected $cacheManager;

    public function __construct(CacheManager $cache = null)
    {
        $this->cacheManager = $cache ?: GeneralUtility::makeInstance(CacheManager::class);
    }

    public function __invoke(AfterTcaCompilationEvent $event)
    {
        $this->initializeColPosCache($event->getTca());
    }

    public function initializeColPosCache(array $tca): array
    {
        $configuration = [];

        foreach ($tca as $table => $tableConfiguration) {
            if (empty($tableConfiguration['columns'])) {
                continue;
            }

            foreach ($tableConfiguration['columns'] as $field => $fieldConfiguration) {
                if (!$this->isInlineField($fieldConfiguration)
                    || !$this->isValidForeignTable($fieldConfiguration)
                    || !$this->hasColPosConfiguration($fieldConfiguration)
                ) {
                    continue;
                }

                $colPos = (int)$fieldConfiguration['config']['overrideChildTca']['columns']['colPos']['config']['default'];
                if (!isset($configuration[$colPos][$table])) {
                    $configuration[$colPos][$table] = [];
                }
                $configuration[$colPos][$table][] = $field;
            }
        }

        $this->cacheManager->set($configuration);

        return [$tca];
    }

    protected function isInlineField(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['type'])
            && $fieldConfiguration['config']['type'] === 'inline';
    }

    protected function isValidForeignTable(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['foreign_table'])
            && !empty($fieldConfiguration['config']['foreign_field'])
            && $fieldConfiguration['config']['foreign_table'] === 'tt_content';
    }

    protected function hasColPosConfiguration(array $fieldConfiguration): bool
    {
        return !empty($fieldConfiguration['config']['overrideChildTca']['columns']['colPos']['config']['default']);
    }
}


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "hide_used_content".
 *
 * Auto generated 31-08-2023 11:11
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
  'title' => 'Hide used content elements',
  'description' => 'Removes used content elements with own colPos configuration from "Unused" column',
  'category' => 'misc',
  'author' => 'Nicole Cordes',
  'author_email' => 'typo3@cordes.co',
  'author_company' => 'biz-design',
  'state' => 'stable',
  'uploadfolder' => 0,
  'createDirs' => '',
  'clearCacheOnLoad' => 0,
  'version' => '1.1.0',
  'constraints' =>
  array (
    'depends' =>
    array (
      'typo3' => '10.4.0-12.4.99',
    ),
    'conflicts' =>
    array (
    ),
    'suggests' =>
    array (
    ),
  ),
  '_md5_values_when_last_written' => 'a:14:{s:9:"ChangeLog";s:4:"b010";s:7:"LICENSE";s:4:"b234";s:9:"README.md";s:4:"3356";s:13:"composer.json";s:4:"484d";s:13:"composer.lock";s:4:"769e";s:12:"ext_icon.png";s:4:"ad17";s:17:"ext_localconf.php";s:4:"ea19";s:16:"phpunit.xml.dist";s:4:"316a";s:24:"sonar-project.properties";s:4:"3a2b";s:30:"Classes/Cache/CacheManager.php";s:4:"7606";s:48:"Classes/EventListener/TcaColPosEventListener.php";s:4:"73f4";s:36:"Classes/Hooks/PageLayoutViewHook.php";s:4:"91e0";s:27:"Configuration/Services.yaml";s:4:"53f6";s:36:"Resources/Public/Icons/Extension.svg";s:4:"0e56";}',
);

